swe_roe/ex2.c

```diff
@@ -0,0 +1,1224 @@
+static char help[] = "Partial 2D dam break problem.\n";
+
+#include <assert.h>
+#include <math.h>
+#include <petsc.h>
+#include <petscdmda.h>
+#include <petscmat.h>
+#include <petscts.h>
+#include <petscvec.h>
+
+#define RDyAlloc(size, result) PetscCalloc1(size, result)
+
+typedef struct _n_User *User;
+
+typedef struct {
+  PetscReal X[3];
+} RDyCoordinate;
+
+typedef struct {
+  PetscReal V[3];
+} RDyVector;
+
+typedef enum {
+  CELL_TRI_TYPE = 0, /* tetrahedron cell for a 3D cell */
+  CELL_QUAD_TYPE     /* hexahedron cell for a 3D cell */
+} RDyCellType;
+
+typedef struct {
+  PetscInt *id;         /* id of the cell in local numbering */
+  PetscInt *global_id;  /* global id of the cell in local numbering */
+  PetscInt *natural_id; /* natural id of the cell in local numbering */
+
+  PetscBool *is_local;
+
+  PetscInt *num_vertices;  /* number of vertices of the cell    */
+  PetscInt *num_edges;     /* number of edges of the cell       */
+  PetscInt *num_neighbors; /* number of neigbors of the cell    */
+
+  PetscInt *vertex_offset;   /* vertice IDs that form the cell    */
+  PetscInt *edge_offset;     /* vertice IDs that form the cell    */
+  PetscInt *neighbor_offset; /* vertice IDs that form the cell    */
+
+  PetscInt *vertex_ids;   /* vertice IDs that form the cell    */
+  PetscInt *edge_ids;     /* edge IDs that form the cell       */
+  PetscInt *neighbor_ids; /* neighbor IDs that form the cell   */
+
+  RDyCoordinate *centroid; /* cell centroid                     */
+
+  PetscReal *area; /* area of the cell                */
+
+} RDyCell;
```

Comment:
What about RDyCells, RDyVertices, and RDyEdges for these three structs, instead of the singhlar forms? That makes clear that they are arrays of data for several items instead of one. (Not a big deal, but naming is hard, so I think it deserves some thought.)